transport/mixer/src/channel.rs

```diff
@@ -0,0 +1,487 @@
+lazy_static::lazy_static! {
+    pub static ref METRIC_QUEUE_SIZE: SimpleGauge =
+        SimpleGauge::new("hopr_mixer_queue_size", "Current mixer queue size").unwrap();
+    pub static ref METRIC_MIXER_AVERAGE_DELAY: SimpleGauge = SimpleGauge::new(
+        "hopr_mixer_average_packet_delay",
+        "Average mixer packet delay averaged over a packet window"
+    )
+    .unwrap();
+}
+
+/// Mixing and delaying channel using random delay function.
+///
+/// Mixing is performed by assigning random delays to the ingress timestamp of data,
+/// then storing the values inside a binary heap with reversed ordering (max heap).
+/// This effectively creates a min heap behavior, which is required to ensure that
+/// data is released in order of their delay expiration.
+///
+/// When data arrives:
+/// 1. A random delay is assigned
+/// 2. Data is stored in the heap with its release timestamp
+/// 3. The heap maintains ordering so items with earliest release time are at the top
+///
+/// The channel uses a single timer thread that is instantiated on the first
+/// timer reset and shared across all operations. This channel is **unbounded** by nature
+/// using the `capacity` in the configuration to solely pre-allocate the buffer.
+struct Channel<T> {
+    /// Buffer holding the data with a timestamp ordering to ensure the min heap behavior.
+    buffer: BinaryHeap<Reverse<DelayedData<T>>>,
+    timer: futures_timer::Delay,
+    waker: Option<std::task::Waker>,
+    cfg: MixerConfig,
+}
+
+/// Channel with sender and receiver counters allowing closure tracking.
+struct TrackedChannel<T> {
+    channel: Arc<Mutex<Channel<T>>>,
+    sender_count: Arc<AtomicUsize>,
+    receiver_active: Arc<AtomicBool>,
+}
+
+impl<T> Clone for TrackedChannel<T> {
+    fn clone(&self) -> Self {
+        Self {
+            channel: self.channel.clone(),
+            sender_count: self.sender_count.clone(),
+            receiver_active: self.receiver_active.clone(),
+        }
+    }
+}
+
+/// Error returned by the [`Sender`].
+#[derive(Debug, thiserror::Error)]
+pub enum SenderError {
+    /// The channel is closed due to receiver being dropped.
+    #[error("Channel is closed")]
+    Closed,
+
+    /// The mutex lock over the channel failed.
+    #[error("Channel lock failed")]
+    Lock,
+}
+
+/// Sender object interacting with the mixing channel.
+pub struct Sender<T> {
+    channel: TrackedChannel<T>,
+}
+
+impl<T> Clone for Sender<T> {
+    fn clone(&self) -> Self {
+        let channel = self.channel.clone();
+        channel.sender_count.fetch_add(1, Ordering::Relaxed);
+
+        Sender { channel }
+    }
+}
+
+impl<T> Drop for Sender<T> {
+    fn drop(&mut self) {
+        self.channel.sender_count.fetch_sub(1, Ordering::Relaxed);
+        if self.channel.sender_count.load(Ordering::Relaxed) == 0
+            && !self.channel.receiver_active.load(Ordering::Relaxed)
+        {
+            let channel = self.channel.channel.lock();
```

Comment:
At this point you can already `clear_poison` (and safely `unwrap` thereafter) since the channel is closing down anyway, because we're dropping the last active sender.